### ISSUE TITLE
FinalLocalVariable: case fall-through false positive

### ISSUE THREAD

**Issue by Luolc** [checkstyle/checkstyle#4224]
Taken from https://github.com/checkstyle/checkstyle/pull/4115#discussion_r110142017

```
$ javac Foo.java
$ cat Foo.java
public class Foo {
    void foo() {
        int a; // expect no violation
        final int random = (int) Math.random();
        switch (random) {
            case 0:
                a = 1;
            default:
                a = 2;
                break;
        }
    }
}
```

```
$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalLocalVariable">
        </module>
    </module>
</module>
```

```
$ java -jar checkstyle-7.7-SNAPSHOT-all.jar -c config.xml Foo.java
Starting audit...
[ERROR] /Users/LuoLiangchen/personal/develop/java/checkstyle/reproduction/Foo.java:3:13: 变量 'a' 应被声明为 final 的。 [FinalLocalVariable]
Audit done.
Checkstyle ends with 1 errors.
```

Expected: no violations.

**Comment by Gaurav-Punjabi** [comment]
I'm on it

**Comment by stoyanK7** [comment]
I am on it.

**Comment by stoyanK7** [comment]
Laying out my thoughts on how this should be approached.

Multiple `CASE_GROUP`s should be considered as a single scope until there is one that contains a `LITERAL_BREAK` or the last one is the last one in the switch statement. A `LITERAL_BREAK` should be the end of the scope.
```java
switch (x) {
    case 0:
        a = 1;
    case 1:
        a = 2;
        break; // case 0 and 1 are a single scope
    case 2:
        a = 3;
        break; // this is another scope
    case 3:
        a = 4;
    case 4:
        a = 5;
    case 6:
        a = 7;
        break; // case 3,4 and 5 are a single scope
}

switch (x) {
    case 0:
        a = 1;
    case 1:
        a = 2;
    case 2:
        a = 3;  // all the cases are a single scope
}
```

There is also an edge case that needs to be taken into consideration. Following the logic above, the 2 `CASE_GROUP`s below would be 2 different scopes, but Java does not see it that way.
https://github.com/checkstyle/checkstyle/blob/1346ff94acef40b7772b5170beeb5fbc5d5b7d5a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariable.java#L281-L289

**Comment by romani** [comment]
as first step we can remove false positive (pain for users and might be build failure on their side) and allow to have false negative (violation should be but not present)
and as second step try to resolve false-negatives with some/all details on scopes and presence in all case and ...
